django/db/models/query_utils.py

```diff
@@ -310,6 +311,10 @@ def merge_dicts(dicts):
             merged.update(d)
         return merged
 
+    @staticmethod
+    def remove_none_from_dict(dict_):
+        return {k: v for k, v in dict_.items() if v is not None}
+
```

Comment:
What breaks exactly when `None` are kept in there?

I think they are important to denote unregistration particularly when merging instance lookups with class lookups. If you prune them the _unregistered_ sentinel from instance lookups goes away which explains why `test_unregister_instance_lookup` fails.

Reply:
Look at commits 2 & 3 where I restored the behavior of the existing tests.

In one case where the order of actions is "register A on class, register B on instance, unregister B on instance"; the new behavior would be that A is no longer registered on the instance because it's overridden by B's "None" sentinel.

Reply:
I see.

Shouldn't we differentiate instance unregistration to pop the key when it's currently registered at an instance level and assign `None` otherwise instead?